### PR TITLE
release-23.1.20-rc: stats: more conservative lower bound for downward-trending forecasts

### PR DESCRIPTION
**Pull request by michae2**
Backport:
  * 1/1 commits from "stats: more conservative lower bound for downward-trending forecasts" (#122273)
  * 1/1 commits from "sql/stats: turn forecasting constants into cluster settings" (#122779)

/cc @cockroachdb/release

---

**stats: more conservative lower bound for downward-trending forecasts**

Instead of 0 as a lower bound for predictions, use 1/3 times the lowest prior observation. This will avoid predicting zero rows for a statistic unless we really observed zero rows somewhat recently.

Fixes: #119967

Release note (bug fix): Statistics forecasts of zero rows can cause bad plans. This commit changes forecasting to avoid predicting zero rows for most downward-trending statistics.

**sql/stats: turn forecasting constants into cluster settings**

Turn three constants used for statistics forecasting into cluster settings. (These must be cluster settings rather than session variables because forecasting happens when reloading table statistics in the stats cache, which can be independent of any session.)

Informs: #119967

Release note (sql change): Introduce three new cluster settings for controlling table statistics forecasting:

1. `sql.stats.forecasts.min_observations` is the minimum number of observed statistics required to produce a forecast.

2. `sql.stats.forecasts.min_goodness_of_fit` is the minimum R² (goodness of fit) measurement required from all predictive models to use a forecast.

3. `sql.stats.forecasts.max_decrease` is the most a prediction can decrease, expressed as the minimum ratio of the prediction to the lowest prior observation.

---

Release justification: specially-requested fix for plan regression bug.

### CODE REVIEW

**Review by yuzefovich — APPROVED** [review]
> Reviewed 3 of 3 files at r1, 4 of 4 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/122990)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @mgartner)

**Comment by michae2** [comment]
Fixes for the tests coming up, but this is RFAL.

**Review by mgartner — APPROVED** [review]
:lgtm:

> Reviewed 3 of 3 files at r1, 4 of 4 files at r3, 5 of 5 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/122990#-:-NwHcuXbB6cg_aS3QUc2:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @michae2)

File: pkg/sql/stats/forecast.go [hunk 1]
```diff
@@ -44,11 +44,58 @@ var UseStatisticsForecasts = settings.RegisterBoolSetting(
...
+// cause bad plans when the forecast is initially used.
+//
+// This happens to be the same as unknownFilterSelectivity, but there's not a
+// strong theoretical reason for it.
+var maxDecrease = settings.RegisterFloatSetting(
+	settings.TenantWritable,
+	"sql.stats.forecasts.max_decrease",
+	"the most a prediction is allowed to decrease, expressed as the minimum ratio of the prediction "+
+		"to the lowest prior observation",
+	1.0/3.0,
```
Comment (mgartner) [inline pkg/sql/stats/forecast.go:89]: Should we set this to 0 be default in backports? Would that reduce risk of this change?

Comment (michae2) [reply pkg/sql/stats/forecast.go:89]: Good point, I will do that.

**Review by rytaft — APPROVED** [review]
:lgtm:

> Reviewed 3 of 3 files at r1, 4 of 4 files at r3, 5 of 5 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/122990#-:-NwKscGJ7aRpjgYrEvnB:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @mgartner)

**Pull request merged**